juju/terraform-provider-juju
# Implement endpoint bindings for applications (#400)

## Description

The goal is to enable binding endpoints to specific spaces. This features will used in sunbeam when deploying with MAAS.

Fixes: #400 

## Type of change

*\<What type of a change is this? Please keep one or more relevant lines from below and delete the others.\>*

- Change existing resource
- Logic changes in resources (the API interaction with Juju has been changed)
- Change in tests (one or several tests have been changed)
- Requires a documentation update
- Maintenance work (repository related, like Github actions, or revving the Go version, etc.)

## Environment

- Juju controller version: 3.4

- Terraform version: 1.6

## QA steps

Manual QA steps should be done to test this PR.

- You need to have the spaces `management` and `public` associated to the model
```bash
lxc network create management-br ipv4.address=10.150.40.1/24 ipv4.nat=true ipv6.address=none ipv6.nat=false
lxc network create public-br ipv4.address=10.170.80.1/24 ipv4.nat=true ipv6.address=none ipv6.nat=false
juju add-model my-model
juju add-space management 10.150.40.0/24
juju add-space public 10.170.80.0/24
```

```tf
provider juju {}

data "juju_model" "resource" {
  name = "my-model"
}

resource "juju_application" "zk" {
  name  = "zookeeper"
  model = data.juju_model.resource.name

  constraints = "arch=amd64 spaces=management,public"
  charm {
    name    = "zookeeper"
    channel = "3/stable"
    base    = "ubuntu@22.04"
  }

  endpoint_bindings = [
    {
      space = "public"
    },
    {
      endpoint = "zookeeper"
      space    = "management"
    }
  ]
}
```

## Additional notes

*\<Please add relevant notes & information, links to mattermost chats, other related issues/PRs, anything to help understand and QA the PR.\>*
Some tests were added, with a note about changing space setup when #336 is implemented.
Current integration tests have been updated to add 2 lxd bridges as support for the tests.
Tests need `TEST_PUBLIC_BR` and `TEST_MANAGEMENT_BR` containing the cidrs to be run, otherwise, they are skipped.

On behavior:
Specifying a space without an endpoint will make the space the default space (in term of bindings, not touching model configs)

Constraints are in QA/tests to make sure LXD brings the instance with right networking configuration.
